modules/contact/src/userobjects/PenaltyWeightedGapUserObject.C

```diff
@@ -78,15 +114,82 @@ PenaltyWeightedGapUserObject::reinit()
   for (const auto i : make_range(_test->size()))
   {
     const Node * const node = _lower_secondary_elem->node_ptr(i);
-    const auto & weighted_gap =
-        libmesh_map_find(_dof_to_weighted_gap, static_cast<const DofObject *>(node)).first;
-    const auto weighted_gap_for_calc = weighted_gap < 0 ? -weighted_gap : ADReal(0);
+
+    // Simo's definition of the gap is positive during penetration
+    const auto & weighted_gap = -libmesh_map_find(_dof_to_weighted_gap, node).first;
+    const auto lagrange_multiplier =
+        _augmented_lagrange_problem ? _dof_to_lagrange_multiplier[node] : 0.0;
     const auto & test_i = (*_test)[i];
+
+    // Simo et al. 2.14
+    auto normal_pressure = _penalty * weighted_gap + lagrange_multiplier;
+    normal_pressure = normal_pressure > 0.0 ? normal_pressure : 0.0;
+
+    _dof_to_normal_pressure[node] = normal_pressure;
     for (const auto qp : make_range(_qrule_msm->n_points()))
+      _contact_force[qp] += test_i[qp] * normal_pressure;
+  }
+}
+
+void
+PenaltyWeightedGapUserObject::selfTimestepSetup()
+{
+  // do not clear the LMs!
+
+  // predict next time step LM linearly
+  if (_predictor_scale != 0.0)
+    for (auto & [dof_object, lagrange_multiplier] : _dof_to_lagrange_multiplier)
     {
-      _contact_force[qp] += (test_i[qp] * _penalty) * weighted_gap_for_calc;
-      _dof_to_normal_pressure[static_cast<const DofObject *>(node)] =
-          MetaPhysicL::raw_value(_penalty * weighted_gap_for_calc);
+      // save off current LM
+      const auto current_lm = lagrange_multiplier;
+
+      // find old LM
+      const auto it = _dof_to_old_lagrange_multiplier.find(dof_object);
+      if (it != _dof_to_old_lagrange_multiplier.end())
+      {
+        // Linearly extrapolate LM
+        const auto old_lm = it->second;
+        lagrange_multiplier += (current_lm - old_lm) / _dt_old * _dt * _predictor_scale;
+      }
+
+      // save unupdated LM old LM storage
+      _dof_to_old_lagrange_multiplier[dof_object] = current_lm;
+    }
+
+  // save old timestep
+  _dt_old = _dt;
+}
+
+void
+PenaltyWeightedGapUserObject::timestepSetup()
+{
+  selfTimestepSetup();
+}
+
+bool
+PenaltyWeightedGapUserObject::isContactConverged()
+{
+  // check if penetration is below threshold
+  for (const auto & [dof_object, gap] : _dof_to_weighted_gap)
+    if (physicalGap(gap) < -_penetration_tolerance ||
+        (physicalGap(gap) > _penetration_tolerance && _dof_to_lagrange_multiplier[dof_object] > 0))
+      return false;
```

Comment:
It'd be nice to have a machinery to report the least converged node at the end of the time step or maybe also at the end of the nonlinear iteration. Just a thought. Not to be done in this PR.